File: days/02/main.py
import logging
logging.basicConfig(level=logging.DEBUG)

class RangeAnalyzer:
    """
    A Class to analyze ranges defined by start and end points for invalid IDs.
    """
    def __init__(self, logger: logging.Logger):
        self.logger = logger
        self.simple_invalid_id_sum = 0
        self.complicated_invalid_id_sum = 0

    def is_simply_invalid(self, id: int) -> bool:
        """
        Check if the given ID is made only of some sequence of digits repeated twice.
        """
        id_str = str(id)
        length = len(id_str)
        if length % 2 != 0:
            return False
        half = length // 2
        if id_str[:half] == id_str[half:]:
            return True
        return False

    def is_complicated_invalid(self, id: int) -> bool:
        """
        Check if the given ID is made only of some sequence of digits repeated multiple times.
        Works by itterating through each Character and checks if it has been seen in sequence before.
        """
        id_str = str(id)
        
        s = id_str
        n = len(s)

        # Try every possible repeating block length
        for size in range(1, n // 2 + 1):
            if n % size == 0:
                block = s[:size]
                if block * (n // size) == s:
                    return True
        return False


    def analyze_range(self, start: int, end: int) -> None:
        for id in range(start, end + 1):
            if self.is_simply_invalid(id):
                self.simple_invalid_id_sum += id
                self.logger.debug(f"Found simple invalid ID: {id}")
            if self.is_complicated_invalid(id):
                self.complicated_invalid_id_sum += id
                self.logger.debug(f"Found complicated invalid ID: {id}")

    def get_simple_invalid_id_count(self) -> int:
        return self.simple_invalid_id_sum
    
    def get_complicated_invalid_id_sum(self) -> int:
        return self.complicated_invalid_id_sum

def main() -> None:
    logger = logging.getLogger("day_01")
    analyzer = RangeAnalyzer(logger)

    with open("days/02/puzzle_input.txt") as f:
        ranges = f.read().split(",")
        for range in ranges:
            start_str, end_str = range.split("-")
            start, end = int(start_str), int(end_str)
            analyzer.analyze_range(start, end)
    
    print("-"*10 + "SOLUTION DAY 02" + "-"*10)
    print(f"Part1: Easy invalid ID sum {analyzer.get_simple_invalid_id_count()}.")
    print(f"Part2: All invalid ID sum {analyzer.get_complicated_invalid_id_sum()}.")
    print("-"*35)
    return None

if __name__ == "__main__":
    raise RuntimeError("Please run this script through main.py")

File: days/02/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

type RangeAnalyzer struct {
	simple_invalid_id_sum      int
	complicated_invalid_id_sum int
}

func NewRangeAnalyzer() *RangeAnalyzer {
	return &RangeAnalyzer{}
}

// Check if ID is made of 2 repeated halves (e.g., 1212 → "12" + "12")
func (r *RangeAnalyzer) is_simply_invalid(id int) bool {
	id_str := strconv.Itoa(id)
	length := len(id_str)

	if length%2 != 0 {
		return false
	}

	half := length / 2
	return id_str[:half] == id_str[half:]
}

// Check if ID is made of *any* repeating block (e.g., 123123123)
func (r *RangeAnalyzer) is_complicated_invalid(id int) bool {
	id_str := strconv.Itoa(id)
	n := len(id_str)

	for size := 1; size <= n/2; size++ {
		if n%size == 0 {
			block := id_str[:size]
			if strings.Repeat(block, n/size) == id_str {
				return true
			}
		}
	}
	return false
}

func (r *RangeAnalyzer) analyze_range(start, end int) {
	for id := start; id <= end; id++ {
		if r.is_simply_invalid(id) {
			r.simple_invalid_id_sum += id
		}
		if r.is_complicated_invalid(id) {
			r.complicated_invalid_id_sum += id
		}
	}
}

func (r *RangeAnalyzer) get_simple_invalid_id_count() int {
	return r.simple_invalid_id_sum
}

func (r *RangeAnalyzer) get_complicated_invalid_id_sum() int {
	return r.complicated_invalid_id_sum
}

func main() {
	data, err := os.ReadFile("days/02/puzzle_input.txt")
	if err != nil {
		panic(err)
	}

	analyzer := NewRangeAnalyzer()

	ranges := strings.SplitSeq(strings.TrimSpace(string(data)), ",")
	for rng := range ranges {
		parts := strings.Split(rng, "-")

		start, err1 := strconv.Atoi(parts[0])
		end, err2 := strconv.Atoi(parts[1])

		if err1 != nil || err2 != nil {
			panic("Invalid range in input file")
		}

		analyzer.analyze_range(start, end)
	}

	fmt.Println("----------SOLUTION DAY 02----------")
	fmt.Printf("Part1: Easy invalid ID sum %d.\n", analyzer.get_simple_invalid_id_count())
	fmt.Printf("Part2: All invalid ID sum %d.\n", analyzer.get_complicated_invalid_id_sum())
	fmt.Println("-----------------------------------")
}
